### PR TITLE
[Feature] OAuth2 - OIDC Custom implementation

### PR DESCRIPTION
**Pull request by abelgardep**
Implements https://github.com/owncloud/android/issues/3038
Needed for https://github.com/owncloud/android/pull/3043

### CODE REVIEW

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/GetOIDCDiscoveryRemoteOperation.kt [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+                // Parse the response
+                val moshi: Moshi = Moshi.Builder().build()
+                val jsonAdapter: JsonAdapter<OIDCDiscoveryResponse> = moshi.adapter(OIDCDiscoveryResponse::class.java)
+                val oidcDiscoveryResponse: OIDCDiscoveryResponse? = jsonAdapter.fromJson(responseBody)
+
+                result = RemoteOperationResult(RemoteOperationResult.ResultCode.OK)
+                result.data = oidcDiscoveryResponse
+
+                Timber.d("Get OIDC Discovery completed and parsed to $oidcDiscoveryResponse")
```
Comment (theScrabi) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/GetOIDCDiscoveryRemoteOperation.kt:75]:
```suggestion
                val result = RemoteOperationResult(RemoteOperationResult.ResultCode.OK)
                result.data = oidcDiscoveryResponse
                Timber.d("Get OIDC Discovery completed and parsed to $oidcDiscoveryResponse")
                return result
```

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/GetOIDCDiscoveryRemoteOperation.kt [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            } else {
+                result = RemoteOperationResult(getMethod)
+                Timber.e("Failed response while getting OIDC server discovery from the server status code: $status; response message: $responseBody")
```
Comment (theScrabi) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/GetOIDCDiscoveryRemoteOperation.kt:78]:
```suggestion
                Timber.e("Failed response while getting OIDC server discovery from the server status code: $status; response message: $responseBody")
                return RemoteOperationResult(getMethod)
```

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/GetOIDCDiscoveryRemoteOperation.kt [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            }
+
+        } catch (e: Exception) {
+            result = RemoteOperationResult(e)
+            Timber.e(e, "Exception while getting OIDC server discovery")
```
Comment (theScrabi) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/GetOIDCDiscoveryRemoteOperation.kt:83]:
```suggestion
            Timber.e(e, "Exception while getting OIDC server discovery")
            return RemoteOperationResult(e)
```

**Review by theScrabi — COMMENTED** [review]
I'd opt for a more functional approach whenever possible.

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/TokenRequestRemoteOperation.kt [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+            if (status == HTTP_OK && responseBody != null) {
+                Timber.d("Successful response $responseBody")
+
+                // Parse the response
+                val moshi: Moshi = Moshi.Builder().build()
+                val jsonAdapter: JsonAdapter<TokenResponse> = moshi.adapter(TokenResponse::class.java)
+                val tokenResponse: TokenResponse? = jsonAdapter.fromJson(responseBody)
+
+                result = RemoteOperationResult(RemoteOperationResult.ResultCode.OK)
```
Comment (theScrabi) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/TokenRequestRemoteOperation.kt:80]: Same here.

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/GetOIDCDiscoveryRemoteOperation.kt (hunk 3 above)
Comment (abelgardep) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/GetOIDCDiscoveryRemoteOperation.kt:83]: I see your point. I will update it, thanks.

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/TokenRequestRemoteOperation.kt [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                Timber.d("Get tokens completed and parsed to $tokenResponse")
+
+                return RemoteOperationResult<TokenResponse>(RemoteOperationResult.ResultCode.OK).apply {
+                    data = tokenResponse
+                }
+
+            } else {
+                Timber.e("Failed response while getting tokens from the server status code: $status; response message: $responseBody")
+                return RemoteOperationResult<TokenResponse>(postMethod)
+            }
```
Comment (davigonz) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/TokenRequestRemoteOperation.kt:79]:
```suggestion
with(tokenRequestParams) {
    PostMethod(
        url = URL(tokenEndpoint),
        postRequestBody = toRequestBody()
    ).apply {
        addRequestHeader(AUTHORIZATION_HEADER, clientAuth)
    }.run {
        val status = client.executeHttpMethod(this)
        val responseBody = getResponseBodyAsString()

        return if (status == HTTP_OK && responseBody != null) {
            Timber.d("Successful response $responseBody")
            
            // Parse the response
            val moshi: Moshi = Moshi.Builder().build()
            val jsonAdapter: JsonAdapter<TokenResponse> = moshi.adapter(TokenResponse::class.java)
            val tokenResponse: TokenResponse? = jsonAdapter.fromJson(responseBody)
            Timber.d("Get tokens completed and parsed to $tokenResponse")

           RemoteOperationResult<TokenResponse>(RemoteOperationResult.ResultCode.OK).apply {
               data = tokenResponse
           }
       } else {
           Timber.e("Failed response while getting tokens from the server status code: $status; response message: $responseBody")
            RemoteOperationResult(this)
        }
    }
}
```

Comment (davigonz) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/TokenRequestRemoteOperation.kt:79]: More functional 🍻

**Review by davigonz — CHANGES_REQUESTED** [review]

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/TokenRequestRemoteOperation.kt (hunk 5 above)
Comment (abelgardep) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/TokenRequestRemoteOperation.kt:79]: Yes, I agree that it is more functional, but sometimes I prefer readability. I think it is easier to read this way.

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/HttpConstants.java [hunk 6]
```diff
@@ -51,6 +51,12 @@
     public static final String ACCEPT_ENCODING_IDENTITY = "identity";
     public static final String OC_FILE_REMOTE_ID = "OC-FileId";
 
+    // OAuth
+    public static final String HEADER_AUTHORIZATION_CODE = "code";
+    public static final String HEADER_GRANT_TYPE = "grant_type";
+    public static final String HEADER_REDIRECT_URI = "redirect_uri";
+    public static final String HEADER_REFRESH_TOKEN = "refresh_token";
```
[the remark this answers is not in the document]
Comment (abelgardep) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/HttpConstants.java:58]: I was thinking about that, and finally, I didn't add the `OAuth` prefix. But I agree, I will rename it. Thanks

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/TokenRequestRemoteOperation.kt (hunk 5 above)
Comment (davigonz) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/TokenRequestRemoteOperation.kt:79]: Yep, maybe the `with(tokenRequestParams) {` makes it less readable but without that I like how the post method execution looks like

```
PostMethod(
        url = URL(tokenRequestParams tokenEndpoint),
        postRequestBody = tokenRequestParams.toRequestBody()
    ).apply {
        addRequestHeader(AUTHORIZATION_HEADER, clientAuth)
    }.run {
```

But of course, it's a suggestion, I try to improve my code by reducing immutable variables and transforming them to functional, your way is fine as well.

**Review by davigonz — APPROVED** [review]

**Pull request merged**